File: Tests/VideoStory/Interactor/VideoStoryInteractorTests.m
//
//  VideoStoryInteractorTests.m
//  QBRTCDemo
//
//  Created by Anton Sokolchenko on 20/01/2016.
//  Copyright 2016 Anton Sokolchenko. All rights reserved.
//

#import "BaseTestCase.h"
#import <OCMock/OCMock.h>
#import <OCMStubRecorder.h>
#import <TyphoonPatcher.h>
#import "CallService.h"

#import "FakeCallService.h"
#import "FakeSignalingChannel.h"

#import "VideoStoryInteractor.h"

#import "VideoStoryInteractorOutput.h"

@interface CallService()

@property (nonatomic, assign, readwrite) CallClientState state;

@end

@interface VideoStoryInteractorTests : BaseTestCase

@property (strong, nonatomic) VideoStoryInteractor *interactor;

@property (strong, nonatomic) id mockOutput;

@end

@implementation VideoStoryInteractorTests

#pragma mark - Настройка окружения для тестирования

- (void)setUp {
    [super setUp];
	
    self.interactor = [[VideoStoryInteractor alloc] init];
	
    self.mockOutput = OCMProtocolMock(@protocol(VideoStoryInteractorOutput));
	
    self.interactor.output = self.mockOutput;
	
}

// Use when using real CallService is not possible
- (void)useFakeCallService {
	if ([self.interactor.callService isKindOfClass:[FakeCallService class]]) {
		return; // already set
	}
	self.interactor.callService = [[FakeCallService alloc] initWithSignalingChannel:[FakeSignalingChannel new] clientDelegate:self.interactor];
}

- (void)useRealCallService {
	if ([self.interactor.callService isKindOfClass:[CallService class]]) {
		return; // already set
	}
	self.interactor.callService = [[CallService alloc] initWithSignalingChannel:[FakeSignalingChannel new] clientDelegate:self.interactor];
}

- (void)tearDown {
    self.interactor = nil;

    self.mockOutput = nil;

    [super tearDown];
}

#pragma mark - Тестирование методов VideoStoryInteractorInput

- (void)testConnectingWithUser1 {
	// given
	[self useRealCallService];
	// when
	[self.interactor connectToChatWithUser1];
	
	// then
	OCMVerify([self.mockOutput didConnectToChatWithUser1]);
}

- (void)testConnectingWithUser2 {
	// given
	[self useRealCallService];
	// when
	[self.interactor connectToChatWithUser2];
	
	// then
	OCMVerify([self.mockOutput didConnectToChatWithUser2]);
}

- (void)testHangup {
	// given
	[self useRealCallService];
	
	((CallService *) self.interactor.callService).state = kClientStateConnected;
	
	// when
	[self.interactor hangup];
	
	// then
	OCMVerify([self.mockOutput didHangup]);
}

- (void)testSuccessfulSetLocalCaptureSession {
	// given
	[self useRealCallService];
	
	id mockedInteractor = OCMPartialMock(self.interactor);
	
	OCMStub([mockedInteractor client:[OCMArg any] didReceiveLocalVideoTrack:[OCMArg any]]).andCall(self.interactor.output, @selector(didSetLocalCaptureSession:));
	
	OCMExpect([self.mockOutput didSetLocalCaptureSession:[OCMArg any]]);
	
	// when
	[mockedInteractor connectToChatWithUser1];
	[mockedInteractor startCall];
	
	// then
	OCMVerifyAllWithDelay(self.mockOutput, 10);
}

- (void)testReceiveRemoteVideoTrack_whenConnectedAndStartedCall {
	// given
	[self useFakeCallService];
	
	// when
	[self.interactor connectToChatWithUser1];
	[self.interactor startCall];
	
	// then
	OCMVerify([self.mockOutput didReceiveRemoteVideoTrackWithConfigurationBlock:[OCMArg any]]);
}


@end

File: Tests/VideoStory/Interactor/VideoStoryInteractorTests.swift
//
//  VideoStoryInteractorTests.swift
//  QBRTCDemo
//
//  Created by Anton Sokolchenko on 30.11.16.
//  Copyright © 2016 Anton Sokolchenko. All rights reserved.
//

import XCTest

#if QBRTCDemo_s
	@testable
	import QBRTCDemo_s
#elseif QBRTCDemo
	@testable
	import QBRTCDemo
#endif


// Note: when possible, useRealCallService is used
class VideoStoryInteractorTests: XCTestCase {

	var interactor: VideoStoryInteractor!
	var mockOutput: MockPresenter!
	var testUser: SVUser!
	var testUser2: SVUser!
	
	override func setUp() {
		super.setUp()
		testUser = TestsStorage.svuserRealUser1()
		testUser2 = TestsStorage.svuserRealUser2()
		interactor = VideoStoryInteractor()
		mockOutput = MockPresenter()
		interactor.output = mockOutput
	}
	
	override func tearDown() {
		interactor.output = nil
		interactor = nil
		mockOutput = nil
		super.tearDown()
	}
	
	func useFakeCallService() {
		interactor.callService = FakeCallService(signalingChannel: FakeSignalingChannel(), callServiceDelegate: interactor, dataChannelDelegate: interactor)
	}
	
	func useRealCallService() {
		interactor.callService = CallService(signalingChannel: FakeSignalingChannel(), callServiceDelegate: interactor, dataChannelDelegate: interactor)
	}
	
	// MARK: - Testing methods VideoStoryInteractorInput
	
	func testConnectsWithTestUser() {
		// given
		useRealCallService()
		let testUser = TestsStorage.svuserTest()
		
		// when
		interactor.connectToChatWithUser(testUser, callOpponent: nil)
		
		// then
		XCTAssertEqual(testUser, mockOutput.connectedToChatUser)
		XCTAssertTrue(mockOutput.didConnectToChatWithUserGotCalled)
	}
	
	func testHangup() {
		// given
		useRealCallService()
		
		let callService = interactor.callService
		callService.state = CallServiceState.ClientStateConnected
		
		// when
		interactor.hangup()
		
		// then
		XCTAssertTrue(mockOutput.didHangupGotCalled)
	}
	
	func testSuccessfullySetsLocalCaptureSession() {
		// given
		useRealCallService()
		
		// when
		interactor.connectToChatWithUser(testUser, callOpponent: testUser2)
		
		// then
		XCTAssertTrue(mockOutput.didSetLocalCaptureSessionGotCalled)
	}
	
	func testReceivesRemoteVideoTrack_whenConnectedAndStartedCall() {
		// given
		useFakeCallService()
		
		// when
		interactor.connectToChatWithUser(testUser, callOpponent: testUser2)
		
		// then
		XCTAssertTrue(mockOutput.didReceiveRemoteVideoTrackWithConfigurationBlockGotCalled)
	}
	
	// MARK:- Data Channel Tests
	func testTriggersDidOpenDataChannelOpen_whenReceivedDataChannel() {
		// given
		useFakeCallService()
		
		// when
		interactor.callService.dataChannelEnabled = true
		interactor.connectToChatWithUser(testUser, callOpponent: testUser2)
		
		// then
		XCTAssertTrue(mockOutput.didOpenDataChannelGotCalled)
	}
	
	func testTriggersDidReceiveDataChannelStateReady_whenReceivedDataChannel() {
		// given
		useFakeCallService()
		
		// when
		interactor.callService.dataChannelEnabled = true
		interactor.connectToChatWithUser(testUser, callOpponent: testUser2)
		interactor.requestDataChannelState()
		
		// then
		XCTAssertTrue(mockOutput.didReceiveDataChannelStateReadyGotCalled)
	}
	
	func testTriggersDidReceiveDataChannelStateNotReady_whenNoDataChannel() {
		// given
		useRealCallService()
		
		// when
		interactor.callService.dataChannelEnabled = true
		interactor.requestDataChannelState()
		
		// then
		
		XCTAssertTrue(mockOutput.didReceiveDataChannelStateNotReadyGotCalled)
	}
	
	func testTriggersDidReceiveInvitationToOpenImageGallery_whenReceivedDataChannelInvitationMessage() {
		// given
		useFakeCallService()
		
		// when
		interactor.callService.dataChannelEnabled = true
		interactor.connectToChatWithUser(testUser, callOpponent: testUser2)
		
		let invitationData = DataChannelMessages.invitationToOpenImageGallery().dataUsingEncoding(NSUTF8StringEncoding)
		let invitationMessage = RTCDataBuffer(data: invitationData, isBinary: false)
		
		interactor.callService.channel(nil, didReceiveMessageWithBuffer: invitationMessage)
		
		// then
		XCTAssertTrue(mockOutput.didReceiveInvitationToOpenImageGalleryGotCalled)
	}
	
	func testTriggersSendInvitationToOpenImageGallery_whenRequestedImageGallery() {
		// given
		useFakeCallService()
		
		// when
		interactor.callService.dataChannelEnabled = true
		interactor.connectToChatWithUser(testUser, callOpponent: testUser2)
		
		guard let fakeCallService = interactor.callService as? FakeCallService else {
			XCTFail("We have to use fakeCallService here")
			return
		}
		
		fakeCallService.setHasActiveCall(true)
		
		interactor.sendInvitationMessageAndOpenImageGallery()
		
		// then
		XCTAssertTrue(mockOutput.didSendInvitationToOpenImageGalleryGotCalled)
	}
	
	
	class MockPresenter: NSObject,  VideoStoryInteractorOutput {
		var didConnectToChatWithUserGotCalled = false
		var connectedToChatUser: SVUser?
		
		var didHangupGotCalled = false
		var didFailToConnectToChatGotCalled = false
		var didSetLocalCaptureSessionGotCalled = false
		var didReceiveRemoteVideoTrackWithConfigurationBlockGotCalled = false
		var didOpenDataChannelGotCalled = false
		var didReceiveDataChannelStateReadyGotCalled = false
		var didReceiveDataChannelStateNotReadyGotCalled = false
		var didReceiveInvitationToOpenImageGalleryGotCalled = false
		var didSendInvitationToOpenImageGalleryGotCalled = false
		
		func didConnectToChatWithUser(user: SVUser) {
			connectedToChatUser = user
			didConnectToChatWithUserGotCalled = true
		}
		
		func didHangup() {
			didHangupGotCalled = true
		}
		
		func didFailToConnectToChat() {
			
		}
		
		func didSetLocalCaptureSession(localCaptureSession: AVCaptureSession) {
			didSetLocalCaptureSessionGotCalled = true
		}
		
		func didReceiveRemoteVideoTrackWithConfigurationBlock(block: ((renderer: RTCEAGLVideoView?) -> Void)?) {
			didReceiveRemoteVideoTrackWithConfigurationBlockGotCalled = true
		}
		func didOpenDataChannel() {
			didOpenDataChannelGotCalled = true
		}
		
		func didReceiveDataChannelStateReady() {
			didReceiveDataChannelStateReadyGotCalled = true
		}
		
		func didReceiveDataChannelStateNotReady() {
			didReceiveDataChannelStateNotReadyGotCalled = true
		}
		
		func didReceiveInvitationToOpenImageGallery() {
			didReceiveInvitationToOpenImageGalleryGotCalled = true
		}
		
		func didSendInvitationToOpenImageGallery() {
			didSendInvitationToOpenImageGalleryGotCalled = true
		}
	}
	
}
